from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

# save_flag
gen.add("save_flag", bool_t, 0, "Flag to save the configuration file", True)

# patchworkpp
gen.add("sensor_height", double_t, 0, "Height of the sensor from the ground", 1.723, 0.0, 100.0)
gen.add("mode", str_t, 0, "Mode of operation", "czm")
gen.add("verbose", bool_t, 0, "Verbose mode for debugging", False)
gen.add("visualize", bool_t, 0, "Flag to visualize the ground likelihood estimation", True)

# Ground Plane Fitting parameters
gen.add("num_iter", int_t, 0, "Number of iterations for ground plane estimation using PCA", 3, 0, 10)
gen.add("num_lpr", int_t, 0, "Maximum number of points to be selected as lowest points representative", 20, 0, 100)
gen.add("num_min_pts", int_t, 0, "Minimum number of points to be estimated as ground plane in each patch", 10, 0, 100)
gen.add("th_seeds", double_t, 0, "Threshold for lowest point representatives using in initial seeds selection of ground points", 0.3, 0.0, 1.0)
gen.add("th_dist", double_t, 0, "Threshold for thickness of ground", 0.125, 0.0, 1.0)
gen.add("th_seeds_v", double_t, 0, "Threshold for lowest point representatives using in initial seeds selection of vertical structural points", 0.25, 0.0, 1.0)
gen.add("th_dist_v", double_t, 0, "Threshold for thickness of vertical structure", 0.1, 0.0, 1.0)
gen.add("max_r", double_t, 0, "Maximum range of ground estimation area", 80.0, 0.0, 1000.0)
gen.add("min_r", double_t, 0, "Minimum range of ground estimation area", 2.7, 0.0, 100.0)
gen.add("uprightness_thr", double_t, 0, "Threshold of uprightness using in Ground Likelihood Estimation(GLE)", 0.707, 0.0, 1.0)
gen.add("adaptive_seed_selection_margin", double_t, 0, "The points below the adaptive_seed_selection_margin * sensor_height are filtered", -1.2, -10.0, 10.0)

# czm
gen.add("czm_num_zones", int_t, 0, "Number of zones in czm mode", 4, 0, 10)
gen.add("czm_num_sectors_each_zone", str_t, 0, "Number of sectors in each zone", "16, 32, 54, 32")
gen.add("czm_num_rings_each_zone", str_t, 0, "Number of rings in each zone", "2, 4, 4, 4")
gen.add("czm_elevation_thresholds", str_t, 0, "Threshold of elevation for each ring using in GLE. Those values are updated adaptively.", "0.0, 0.0, 0.0, 0.0")
gen.add("czm_flatness_thresholds", str_t, 0, "Threshold of flatness for each ring using in GLE. Those values are updated adaptively.", "0.0, 0.0, 0.0, 0.0")

# enable/disable options
gen.add("enable_RNR", bool_t, 0, "Flag to enable/disable RNR", True)
gen.add("enable_RVPF", bool_t, 0, "Flag to enable/disable RVPF", True)
gen.add("enable_TGR", bool_t, 0, "Flag to enable/disable TGR", True)

# RNR parameters
gen.add("RNR_ver_angle_thr", double_t, 0, "Noise points vertical angle threshold. Downward rays of LiDAR are more likely to generate severe noise points.", -15.0, -90.0, 90.0)
gen.add("RNR_intensity_thr", double_t, 0, "Noise points intensity threshold. The reflected points have relatively small intensity than others.", 0.2, 0.0, 1.0)

# Generate the .cfg file
exit(gen.generate("patchworkpp", "demo", "Patchworkpp"))